File: windows/GestorFinances.Tests/Domain/Rules/AutoCategorizer.cs
namespace GestorFinances.Domain.Rules;

public sealed record AutoCategorizeMovement(
    string? Name,
    string? Payee,
    long AmountCents,
    DateOnly Date,
    string AccountId);

public sealed record AutoCategorizeConditions(
    string? TextContains = null,
    long? AmountMinCents = null,
    long? AmountMaxCents = null,
    string? AccountId = null,
    IReadOnlySet<int>? DayOfMonthIn = null);

public sealed record AutoCategorizeAction(
    string? CategoryId = null,
    string? TripId = null);

public sealed record AutoCategorizeRule(
    string Id,
    int Priority,
    string CreatedAt,
    bool Active,
    AutoCategorizeConditions Conditions,
    AutoCategorizeAction Action);

public sealed record AutoCategorizeMatch(
    string RuleId,
    AutoCategorizeAction Action);

public static class AutoCategorizer
{
    public static AutoCategorizeMatch? FindMatch(
        AutoCategorizeMovement movement,
        IEnumerable<AutoCategorizeRule> rules) =>
        rules
            .Where(rule => rule.Active)
            .Where(rule => Matches(movement, rule.Conditions))
            .OrderByDescending(rule => rule.Priority)
            .ThenByDescending(rule => rule.CreatedAt, StringComparer.Ordinal)
            .Select(rule => new AutoCategorizeMatch(rule.Id, rule.Action))
            .FirstOrDefault();

    private static bool Matches(AutoCategorizeMovement movement, AutoCategorizeConditions conditions)
    {
        var searchableText = $"{movement.Name ?? ""} {movement.Payee ?? ""}".ToLowerInvariant();

        if (conditions.TextContains is not null &&
            !searchableText.Contains(conditions.TextContains.ToLowerInvariant(), StringComparison.Ordinal))
        {
            return false;
        }

        if (conditions.AmountMinCents is not null && movement.AmountCents < conditions.AmountMinCents)
        {
            return false;
        }

        if (conditions.AmountMaxCents is not null && movement.AmountCents > conditions.AmountMaxCents)
        {
            return false;
        }

        if (conditions.AccountId is not null && movement.AccountId != conditions.AccountId)
        {
            return false;
        }

        if (conditions.DayOfMonthIn is { Count: > 0 } && !conditions.DayOfMonthIn.Contains(movement.Date.Day))
        {
            return false;
        }

        return true;
    }
}

File: android/app/src/main/java/com/gestorfinances/app/domain/rules/AutoCategorizer.kt
package com.gestorfinances.app.domain.rules

import java.time.LocalDate
import java.util.Locale

data class AutoCategorizeMovement(
    val name: String?,
    val payee: String?,
    val amountCents: Long,
    val date: LocalDate,
    val accountId: String,
)

data class AutoCategorizeConditions(
    val textContains: String? = null,
    val amountMinCents: Long? = null,
    val amountMaxCents: Long? = null,
    val accountId: String? = null,
    val dayOfMonthIn: Set<Int> = emptySet(),
)

data class AutoCategorizeAction(
    val categoryId: String? = null,
    val tripId: String? = null,
)

data class AutoCategorizeRule(
    val id: String,
    val priority: Int,
    val createdAt: String,
    val active: Boolean,
    val conditions: AutoCategorizeConditions,
    val action: AutoCategorizeAction,
)

data class AutoCategorizeMatch(
    val ruleId: String,
    val action: AutoCategorizeAction,
)

object AutoCategorizer {
    fun findMatch(
        movement: AutoCategorizeMovement,
        rules: List<AutoCategorizeRule>,
    ): AutoCategorizeMatch? =
        rules
            .asSequence()
            .filter { it.active }
            .filter { matches(movement, it.conditions) }
            .maxWithOrNull(compareBy<AutoCategorizeRule> { it.priority }.thenBy { it.createdAt })
            ?.let { AutoCategorizeMatch(ruleId = it.id, action = it.action) }

    private fun matches(
        movement: AutoCategorizeMovement,
        conditions: AutoCategorizeConditions,
    ): Boolean {
        val searchableText = "${movement.name.orEmpty()} ${movement.payee.orEmpty()}"
            .lowercase(Locale.ROOT)

        if (conditions.textContains != null &&
            !searchableText.contains(conditions.textContains.lowercase(Locale.ROOT))
        ) {
            return false
        }
        if (conditions.amountMinCents != null && movement.amountCents < conditions.amountMinCents) {
            return false
        }
        if (conditions.amountMaxCents != null && movement.amountCents > conditions.amountMaxCents) {
            return false
        }
        if (conditions.accountId != null && movement.accountId != conditions.accountId) {
            return false
        }
        if (conditions.dayOfMonthIn.isNotEmpty() &&
            movement.date.dayOfMonth !in conditions.dayOfMonthIn
        ) {
            return false
        }
        return true
    }
}
